examples/ode_models/ODE_Lotka_Volterra_multiple_ways.myst.md

```diff
@@ -0,0 +1,963 @@
+:category: intermediate, how-to
+:author: [Greg Brunkhorst]{https://github.com/gbrunkhorst}
+:::
+
+```{code-cell} ipython3
+import arviz as az
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import pymc as pm
+import pytensor
+import pytensor.tensor as pt
+
+from pymc.ode import DifferentialEquation
+from pytensor.compile.ops import as_op
+from scipy.integrate import odeint
+from scipy.optimize import least_squares
+
+print(f"Running on PyMC v{pm.__version__}")
+```
+
+```{code-cell} ipython3
+%load_ext watermark
+az.style.use("arviz-darkgrid")
+rng = np.random.default_rng(1234)
+```
+
+## Purpose
+The purpose of this notebook is to demonstrate how to perform Bayesian inference on a system of ordinary differential equations (ODEs), both with and without gradients.  The accuracy and efficiency of different samplers are compared.    
+### Outline
+* Background
+    * Motivation
+    * Lotka-Volterra Predator Prey Model
+    * Hudson's Bay Company Data
+* Scipy `odeint`
+    * Specification
+    * Least Squares Solution
+* Gradient-free Bayesian Inference
+    * Wrapping `odeint` in a Pytensor operator for use in PyMC
+    * Bayesian inference using gradient-free methods
+        * Slice Sampler
+        * DEMetropolisZ Sampler
+        * DEMetropolis Sampler
+        * Metropolis Sampler
+        * Sequential Monte Carlo (SMC) Sampler
+* Bayesian Inference with Gradients
+    * `pymc.ode.DifferentialEquation` specification with the NUTs Sampler
+    * Looping in PyMC with `Pytensor.scan` and the NUTs Sampler
+### Key Conclusions
+* Based on the experiments in this notebook, the most simple and efficient method for performing Bayesian inference on the Lotka-Volterra equations was to specify the ODE system in Scipy, wrap the function as an Pytensor op, and use a Differential Evolution Metropolis (DEMetropolis) sampler in PyMC.  
+
++++ {"tags": []}
+
+## Background
+### Motivation
+Ordinary differential equation models (ODEs) are used in a variety of science and engineering domains to model the time evolution of physical variables.  A natural choice to estimate the values and uncertainty of model parameters given experimental data is Bayesian inference.  However, ODEs can be challenging to specify and solve in the Bayesian setting, therefore, this notebook steps through multiple methods for solving an ODE inference problem using PyMC. The Lotka-Volterra model used in this example has often been used for benchmarking Bayesian inference methods (e.g., in [Stan](https://mc-stan.org/users/documentation/case-studies/lotka-volterra-predator-prey.html)).  See also Richard McElraith's discussion of this model in [Statistical Rethinking](http://xcelab.net/rm/statistical-rethinking/), Chapter 16 of the Second Edition.
+
++++ {"tags": []}
+
+### Lotka-Volterra Predator-Prey Model
+The Lotka-Volterra model describes the interaction between a predator and prey species. This ODE given by:  
+$$
```

Comment:
```suggestion
The Lotka-Volterra model describes the interaction between a predator and prey species. This ODE given by:

$$
```

Try and empty line here, I belive it will fix the rendering:

* current look: https://pymcio--478.org.readthedocs.build/projects/examples/en/478/ode_models/ODE_Lotka_Volterra_multiple_ways.html#lotka-volterra-predator-prey-model
* should look similar to: https://pymcio--478.org.readthedocs.build/projects/examples/en/blog.html/generalized_linear_models/GLM-rolling-regression.html#id1

Reply:
added a line

Reply:
worked!